prereise/gather/griddata/hifld/data_process/demand.py

```diff
@@ -0,0 +1,81 @@
+import inspect
+import os
+
+import pandas as pd
+
+from prereise.gather.griddata import hifld
+from prereise.gather.griddata.hifld import const
+
+
+def assign_demand_to_buses(substations, branch, plant, bus):
+    """Using data on population by county and ZIP code, assign demand to substations,
+    then to the lowest-voltage bus within each substation.
+    This demand parameter is added inplace as a 'Pd' column to the ``bus`` data frame.
+
+    :param pandas.DataFrame substations: table of substation data.
+    :param pandas.DataFrame branch: table of branch data.
+    :param pandas.DataFrame plant: table of plant data.
+    :param pandas.DataFrame bus: table of bus data.
+    """
+    # Load data
+    data_dir = os.path.join(os.path.dirname(inspect.getfile(hifld)), "data")
+    zip_path = os.path.join(data_dir, "uszips.csv")
+    zip_data = pd.read_csv(zip_path, dtype={"zip": "string"}).set_index("zip")
+    county_data = pd.read_csv(os.path.join(data_dir, "uscounties.csv"))
+
+    # Determine each substation's transmission capacity, then sort for selection
+    filtered_branch = branch.query("SUB_1_ID != SUB_2_ID")
+    from_cap = filtered_branch.groupby("SUB_1_ID").sum()["rateA"]
+    to_cap = filtered_branch.groupby("SUB_2_ID").sum()["rateA"]
+    sub_cap = from_cap.combine(to_cap, lambda x, y: x + y, fill_value=0)
+    # Sort substations by their capacities for later ordered selection
+    sorted_subs = substations.loc[sub_cap.sort_values(ascending=False).index].copy()
+
+    # Determine for each ZIP, how much demand to assign to each load substation
+    # Assume here that generator substations don't have load attached to them
+    filtered_subs = sorted_subs.loc[~sorted_subs.index.isin(plant["sub_id"])]
+    subs_per_zip = filtered_subs.value_counts("ZIP")
+    zip_load_substations = subs_per_zip * const.substation_load_share
+    zip_load_substations = zip_load_substations.round().clip(lower=1)
+    zip_assigned_population = (zip_data["population"] / zip_load_substations).dropna()
```

Comment:
When I was reading through the reference code from collaborators, I was thinking whether we should distribute load (population) proportional to the substation capacities instead of uniformly. Do you think that will make a difference?

Reply:
It will definitely make some kind of difference, but I'm not not sure in which general direction the difference will be. Thinking about a low-medium density area, I would imagine that the population is fairly evenly spread out, and so distributing uniformly probably makes sense, since a high-capacity substation may just be a collector, not in response to a pocket of density. However, in an area that truly has different densities, higher-capacity substations may actually be in reaction to higher density. Without further information, I'm not sure what the best conclusion is.

If we run a simulation with this approach and find issues, then I think we may need to revisit this question. I think it's also potentially related to https://github.com/Breakthrough-Energy/PreREISE/issues/234; we may want to patch the algorithm, or we may alternatively want to patch the data that feeds the algorithm.